flow/ThreadHelper.actor.h

```diff
@@ -69,50 +71,110 @@ void onMainThreadVoid(F f, Error* err = nullptr, TaskPriority taskID = TaskPrior
 	}
 
 	virtual void destroy() { UNSTOPPABLE_ASSERT(false); }
 	virtual bool isMultiCallback() const { return false; }
+
+	// MultiCallbackHolder is a helper object for ThreadMultiCallback which allows it to store its index
+	// within the callback vector inside the ThreadCallback rather than having a map of pointers or
+	// some other scheme to store the indices by callback.
+	// MultiCallbackHolder objects can form a doubly linked list.
+	struct MultiCallbackHolder : public FastAllocated<MultiCallbackHolder> {
+		MultiCallbackHolder(ThreadMultiCallback* holder = nullptr,
+		                    MultiCallbackHolder* prev = nullptr,
+		                    MultiCallbackHolder* next = nullptr)
+		  : holder(holder), previous(prev), next(next) {}
+
+		ThreadMultiCallback* holder;
+		int index;
+		MultiCallbackHolder* previous;
+		MultiCallbackHolder* next;
+	};
+
+	// firstHolder is both the inline first record of a MultiCallbackHolder and the head of the
+	// doubly linked list of MultiCallbackHolder entries.
+	MultiCallbackHolder firstHolder;
+
+	// Return a MultiCallbackHolder for the given holder, using the firstHolder if free or allocating
+	// a new one.  No check for an existing record for holder is done.
+	MultiCallbackHolder* addHolder(ThreadMultiCallback* holder) {
+		if (firstHolder.holder == nullptr) {
+			firstHolder.holder = holder;
+			return &firstHolder;
+		}
+		firstHolder.next = new MultiCallbackHolder(holder, &firstHolder, firstHolder.next);
+		return firstHolder.next;
+	}
+
+	// Get the MultiCallbackHolder for holder if it exists, or nullptr.
+	MultiCallbackHolder* getHolder(ThreadMultiCallback* holder) {
+		MultiCallbackHolder* h = &firstHolder;
+		while (h != nullptr && h->holder != holder) {
+			h = h->next;
+		}
+		return h;
+	}
+
+	// Destroy the given MultiCallbackHolder, freeing it if it is not firstHolder.
+	void destroyHolder(MultiCallbackHolder* h) {
+		UNSTOPPABLE_ASSERT(h != nullptr);
+
+		// If h is the firstHolder just clear its holder pointer to indicate unusedness
+		if (h == &firstHolder) {
+			h->holder = nullptr;
+		} else {
+			// Otherwise unlink h from the doubly linked list and free it
+			// h->previous is definitely valid
+			h->previous->next = h->next;
+			if (h->next) {
+				h->next->previous = h->previous;
+			}
+			delete h;
+		}
+	}
 };
 
 class ThreadMultiCallback final : public ThreadCallback, public FastAllocated<ThreadMultiCallback> {
 public:
 	ThreadMultiCallback() {}
 
 	ThreadCallback* addCallback(ThreadCallback* callback) override {
-		UNSTOPPABLE_ASSERT(callbackMap.count(callback) ==
-		                   0); // May be triggered by a waitForAll on a vector with the same future in it more than once
-		callbackMap[callback] = callbacks.size();
+		UNSTOPPABLE_ASSERT(
+		    callback->getHolder(this) ==
+		    nullptr); // May be triggered by a waitForAll on a vector with the same future in it more than once
+		callback->addHolder(this)->index = callbacks.size();
```

Comment:
Since `index` is uninitialized in our holder object, it might be cleaner to require it as an argument to `addHolder` to enforce that it is set. Otherwise, we are requiring the users of `ThreadCallback` to remember to set the index.

Reply:
Done.